waspc/src/Wasp/Generator/WriteFileDrafts.hs

```diff
@@ -0,0 +1,161 @@
+  maybePathsToChecksums <- readChecksumFile dstDir
+  case maybePathsToChecksums of
+    -- If checksums file is missing/corrupted, we delete all of the generated code
+    -- since we are not sure what is the state on the disk and don't want redundant files
+    -- to remain after we write the new file drafts.
+    -- Normally we delete redundant left-over files based on the checksum file, but since we don't have it,
+    -- the only way to be sure is just to delete everything.
+    Nothing -> do
+      removeDirectoryRecursive (SP.fromAbsDir dstDir)
+        `catch` \e -> unless (isDoesNotExistError e) $ throwIO e
+    -- We remove checksum file even if it is valid, so in case we fail in the middle writing on the disk
+    -- and don't get an opportunity to successfully create a new checksum file (reflecting new state on the disk),
+    -- it is clear that something went wrong. Otherwise old checksums file could be accidentally used
+    -- while state on the disk might not be as it says anymore.
+    Just _ -> removeFile $ SP.fromAbsFile $ dstDir </> checksumFileInProjectRoot
+
+  fileDraftsWithChecksums <- mapM (\fd -> (fd,) <$> getChecksum fd) fileDrafts
+
+  let (fileDraftsToWrite, filesToDelete) = fileDraftsToWriteAndFilesToDelete maybePathsToChecksums fileDraftsWithChecksums
+  mapM_ (write dstDir) fileDraftsToWrite
+  deleteFilesAndDirs dstDir filesToDelete
+
+  let relativePathsToChecksums = map (first getDstPath) fileDraftsWithChecksums
+  writeChecksumFile dstDir relativePathsToChecksums
+
+type RelPathsToChecksums = [(FileOrDirPathRelativeTo ProjectRootDir, Checksum)]
+
+-- | This file stores all checksums for files and directories that were written to disk
+-- on the last project generation.
+checksumFileInProjectRoot :: Path' (Rel ProjectRootDir) File'
+checksumFileInProjectRoot = [relfile|.waspchecksums|]
+
+-- | Takes (possibly) existing written file paths and their checksums, and the current
+-- FileDraft and Checksums to be written, and decides which of the new FileDrafts to
+-- write and what redundant files on disk to delete.
+fileDraftsToWriteAndFilesToDelete ::
+  Maybe RelPathsToChecksums ->
+  [(FileDraft, Checksum)] ->
+  ([FileDraft], [FileOrDirPathRelativeTo ProjectRootDir])
+fileDraftsToWriteAndFilesToDelete Nothing fileDraftsWithChecksums =
+  (fst <$> fileDraftsWithChecksums, [])
+fileDraftsToWriteAndFilesToDelete (Just existingFilePathsToChecksums) fileDraftsWithChecksums =
+  let fileDrafts = fst <$> fileDraftsWithChecksums
+   in ( getNewFileDrafts existingFilePathsToChecksums fileDrafts
+          ++ getChangedFileDrafts existingFilePathsToChecksums fileDraftsWithChecksums,
+        getRedundantGeneratedFiles existingFilePathsToChecksums fileDrafts
+      )
+
+getNewFileDrafts :: RelPathsToChecksums -> [FileDraft] -> [FileDraft]
+getNewFileDrafts existingFilePathsToChecksums fileDrafts =
+  filter (\draft -> isNothing $ lookup (getDstPath draft) existingFilePathsToChecksums) fileDrafts
```

Comment:
The lookup looks through the entire array, which may not scale well with a lot of file drafts (linear scaling, making this a nested loop, quadratic scaling). Consider using a Map for existingFilePathsToChecksums and then Map.lookup.

Reply:
Agreed this could become perf issue as project sizes increase.